x/ibc/07-tendermint/types/misbehaviour.go

```diff
@@ -18,38 +18,45 @@ import (
 // of misbehaviour.Header1
 // Similarly, consensusState2 is the trusted consensus state that corresponds
 // to misbehaviour.Header2
-func CheckMisbehaviourAndUpdateState(
-	clientState clientexported.ClientState,
-	consensusState1, consensusState2 clientexported.ConsensusState,
+func (cs *ClientState) CheckMisbehaviourAndUpdateState(
+	ctx sdk.Context,
+	cdc codec.BinaryMarshaler,
+	clientStore sdk.KVStore,
 	misbehaviour clientexported.Misbehaviour,
-	currentTimestamp time.Time,
-	consensusParams *abci.ConsensusParams,
 ) (clientexported.ClientState, error) {
 
-	// cast the interface to specific types before checking for misbehaviour
-	tmClientState, ok := clientState.(*types.ClientState)
-	if !ok {
-		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "expected type %T, got %T", &types.ClientState{}, clientState)
-	}
-
 	// If client is already frozen at earlier height than evidence, return with error
-	if tmClientState.IsFrozen() && tmClientState.FrozenHeight <= uint64(misbehaviour.GetHeight()) {
+	if cs.IsFrozen() && cs.FrozenHeight <= uint64(misbehaviour.GetHeight()) {
 		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidEvidence,
-			"client is already frozen at earlier height %d than misbehaviour height %d", tmClientState.FrozenHeight, misbehaviour.GetHeight())
+			"client is already frozen at earlier height %d than misbehaviour height %d", cs.FrozenHeight, misbehaviour.GetHeight())
 	}
 
-	tmConsensusState1, ok := consensusState1.(*types.ConsensusState)
+	tmEvidence, ok := misbehaviour.(Evidence)
 	if !ok {
-		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "invalid consensus state type for first header: expected type %T, got %T", &types.ConsensusState{}, consensusState1)
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "expected type %T, got %T", misbehaviour, Evidence{})
 	}
-	tmConsensusState2, ok := consensusState2.(*types.ConsensusState)
+
+	// Retrieve trusted consensus states for each Header in misbehaviour
+	// and unmarshal from clientStore
+	consBytes1 := clientStore.Get(host.KeyConsensusState(tmEvidence.Header1.TrustedHeight))
```

Comment:
nit, can you add a few more spacing between lines? it's too packed lol

Reply:
also use `GetConsensusState` helper func

Reply:
I can't because I'm using the `prefix.Store` which only handles `Get` and `Set`, I think this is preferable to giving the specific clients access to the `client.Keeper`

Reply:
see my comment on `update.go`, but the helper function I was referring to is in `07-tendermint/types/store.go` not the client keeper